File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    int height;
    int width;
    
    int dp[210][210];
    int moving[4][2] = { {1,0},{-1,0},{0,1},{0,-1} };
    
    void setDP() {
        
        for (int i = 0; i < height; i++) 
            for (int j = 0; j < width; j++) 
                dp[i][j] = -1;
    }
    
    int find(vector<vector<int>>& matrix, int x, int y) {    
        
        int val = 1;
        
        for (int k = 0; k < 4; k++) {
            
            int nx = x + moving[k][0];
            int ny = y + moving[k][1];
            
            if (nx < 0 || ny < 0 || nx >= height || ny >= width) continue;
            if (matrix[nx][ny] <= matrix[x][y]) continue;
            
            if (dp[nx][ny] > 0) {
                val = max(val, dp[nx][ny] + 1);
            }
            else {
                val = max(val, find(matrix, nx, ny) + 1);
            }
        }
        
        dp[x][y] = val;
        return val;
    }
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        
        height = matrix.size();
        width = matrix[0].size();
        
        setDP();
        
        int ans = 0;
        
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                ans = max(ans, find(matrix, i, j));
            }
        }
    
        return ans;
    }
};

File: 329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.py
class Solution:
    height = 0
    width = 0
    dp = None
    movings = [ [1,0],[-1,0],[0,1],[0,-1] ]
    
    def DFS(self, matrix: List[List[int]], x: int, y: int) -> int:
        
        if self.dp[x][y] > 0:
            return self.dp[x][y]
        
        self.dp[x][y] = 1
        maxDist = 0
        for moving in self.movings:
            nx, ny = x + moving[0], y + moving[1]
            
            if nx < 0 or ny < 0 or nx >= self.height or ny >= self.width:
                continue
            if matrix[nx][ny] <= matrix[x][y]:
                continue
            maxDist = max(maxDist, self.DFS(matrix, nx, ny))
            
        self.dp[x][y] += maxDist
        return self.dp[x][y]
        
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        
        self.height, self.width = len(matrix), len(matrix[0])
        self.dp = [[0 for col in range(self.width)] for row in range(self.height)]
        
        answer = 0
        for i in range(self.height):
            for j in range(self.width):
                answer = max(answer, self.DFS(matrix, i, j))
        return answer
            
        